### ISSUE TITLE
frangi vesselness filter hangs

### ISSUE THREAD

**Issue by kayarre** [vmtk/vmtk#243]
running the command, 
vmtkimagevesselenhancement -ifile DSA.nrrd -method frangi -sigmamin 1.0 -sigmamax 6.0 -sigmasteps 5 -stepmethod 'logarithmic' -scaled 1 -alpha 0.5 -beta 0.5 -gamma 0.5 -ofile DSA_frangi.nrrd

the code appears to hang forever, I killed it after a couple of hours. using the vmtkslicerextension doesn't  hang forever. I tried profiling, but it looks like its spending all the time in waitpid.
unsure how to proceed.

**Comment by rlizzo** [comment]
Thanks Kurt. Can you post details of your system, python version and
whether you're using the latest development Branch?  I haven't been aware
of any issues with the frangi filter, and our tests are passing right now.

I'll look into it tonight.

**Comment by kayarre** [comment]
I can also share the image file, but what's the best way to do that?

**Comment by rlizzo** [comment]
Please if you can. Upload it here. 

https://www.dropbox.com/request/XQBg7fPjBFeFgViQpb0g

**Comment by rlizzo** [comment]
Thanks kurt, I recieved the data file. Will update in a bit.

**Comment by rlizzo** [comment]
So I learned something new today! Apparently vmtkImageVesselEnhancement cannot handle double precision image data. 

I used the following code to debug and correct the problem. While the program didn't hang on my Mac, it only output a 0 leveled image. By converting the data to 'float' (aka vtkImage.Image.GetScalarSize == 4) from 'double' (vtkImage.Image.GetScalarSize == 8), I had no problems, and was able to generate generate the following results on both my mac and linux system (with ubuntu 14.04 LTS installed). 


```python
from vmtk import vmtkscripts

reader = vmtkscripts.vmtkImageReader()
reader.InputFileName = '/Users/rick/projects/vmtk/kurtsansom-DSA.nrrd'
reader.Execute()

print('before casting: ', reader.Image.GetScalarSize())

caster = vmtkscripts.vmtkImageCast()
caster.Image = reader.Image
caster.OutputType = 'float'
caster.Execute()

print('after casting: ', caster.Image.GetScalarSize())

enhancer = vmtkscripts.vmtkImageVesselEnhancement()
enhancer.Image = caster.Image
enhancer.Method = 'frangi'
enhancer.Execute()

viewer = vmtkscripts.vmtkImageViewer()
viewer.Image = enhancer.Image
viewer.Execute()

print('done')
```
printed ->
```
Spacing 0.241920 0.241920 0.241920
Origin -0.000000 -0.000000 -34.271320
Dimensions 256 256 256
before casting:  8
after casting:  4
Quit renderer
done
```

### Resulting Output

![image](https://user-images.githubusercontent.com/17350115/36573511-61ef5dc2-180f-11e8-9d49-3b4ce4f08ecc.png)

Let me know if using imagecast before feeding it into the algorithm works for you. 

@lantiga, 

Any particular reason we are enforcing a `float` type here on line 60? I don't see any reason in the ITK docs that this can't be set to accept a either a float or double type depending on what is fed in?

https://github.com/vmtk/vmtk/blob/463fe5daeda37331a2a1b935c3eb2987d8f6da59/vtkVmtk/Segmentation/vtkvmtkVesselnessMeasureImageFilter.cxx#L58-L70

**Comment by kayarre** [comment]
image cast float and now vmtkvesselenhancement does not hang. 
It would be easy to check for the image type and exit more gracefully or extend the filter to other datatypes?

**Comment by rlizzo** [comment]
Yeah. At the very least we should raise a warning to let the user know that they have an incompatible data type. (Maybe offer to do the conversion for them if running interactively?)

But While it might take a little bit of trial/error to make sure things work, I can't see any obvious reason that we cant operate on double arrays along with float. Int arrays may not play nice, but thats just a hunch (I'd have to get into the ITK docs to verify). If you want to open a PR I'd be happy to review/test it with you. Otherwise Im probably not going to be able to put in a fix until later next week.

In any case, I'm glad to hear things are working for you! Thanks so much for sending in the report!

**Comment by lantiga** [comment]
Hey! Sorry for this hiccup, this thing has been biting a few users in the past.
The only reason why we enforce a float in the filter is ~~laziness~~ ehm, convenience. We could have checked for the pixel type and defined overloads for the method with the appropriate scalar type.

As a matter of facts though, float is ok in my opinion, we don’t need huge numerical precision here to compute the hessian and extract eigenvalues. It wouldn’t hurt to have doubles, but I don’t expect any practical advantage. On the other hand, int images would have to be cast to float anyway.

I propose that we fix the issue by casting the input image to float at the script level, and assert the datatype in the C++ class before passing things to ITK, so we can exit with useful info.

**Comment by kayarre** [comment]
@lantiga In the short term I agree but am also interested in the first proposal as this could apply to several filters.

> We could have checked for the pixel type and defined overloads for the method with the appropriate scalar type.

I have been looking for an example of this , but haven't found a good example. can you point me in the right direction here.

**Comment by lassoan** [comment]
You may find relevant the pattern how 3D Slicer ITK-based modules follow to solve this:

https://github.com/Slicer/Slicer/blob/aa07f6b274778f4cbab626b35228bea8768d38ce/Modules/CLI/GaussianBlurImageFilter/GaussianBlurImageFilter.cxx#L66-L101

It is quite long, not very elegant, but very simple and explicit.

**Comment by lantiga** [comment]
Thanks for the feedback @kayarre, @lassoan's suggestion is the way to go. Right now in most filters in `Segmentation` we just assume that the input image is float in the `SimpleExecute` method, but this is not right.

The way to go is to create a method templated over the type and dispatch using a switch/case from SimpleExecute according to the input type. It should be fairly doable for most filters.

A few scalar types might not be acceptable (e.g. integers most of the times) and we'll need to define what to do, i.e. barf an error or cast to float.

**Comment by kayarre** [comment]
Thank you @lantiga, I really appreciate your time. I created a pull request for a new smoothing filter based on ITK's anisotropic diffusion filter based on something similar to @lassoan suggestion.

#249, and in it use templates to work with different types. Is this implementation approach suitable to be applied to the vesselness filter?
